src/core/backend_context.cc

```diff
@@ -553,7 +624,9 @@ BackendResponder::FlushPendingPinned(
     // response outputs so that we can do the pinned->CPU copies in
```

Comment:
This is in the else clause where a pinned buffer was used. The CopyBuffer above is the one that may be CPU->PINNED and so is the one that needs async if doing CPU->PINNED. If the above did GPU->PINNED then the copy below will always involve GPU and so doesn't need async.